teps/draft-20200716-task-results-in-final-tasks.md

```diff
@@ -0,0 +1,304 @@
+
+**A.** The finally `Task` `boskos-release` is attempted but declared failure since the task result
+`leased-resource` is not initialized. Pipeline controller logs the reason of this failure as invalid result
+reference `InvalidTaskResultReference`. `Pipeline` continues executing rest of the final tasks and exits with failure.
+
+**Q. What happens when the dependent `Task` succeeds but the `Task` `Result` is empty?**
+
+**A.** It could be possible that the task author intentionally chooses to leave the `Task` `Result` empty
+(creating an empty file at `/tekton/results/`). `Pipeline` leaves that param value empty and executes
+the final task. Empty `Task` `Result` is considered a valid value and results in successful variable
+substitution. The error handling of such empty `Task` `Result` is left to the final `Task` author.
+
+**Q. Can a final `Task` produce `Task` `Results`?**
+
+**A.** Nope, this proposal does not enable final tasks to produce `Task` `Results`.
+
+
+**Q. Explain the overall workflow.**
+
+**A.**
+
+```
+                   ------------------------------------------
+                  |             Param Validation              |
+                  | (required params provided and their types)|
+                   ------------------------------------------
+                                    |
+                                    V
+                   -------------------------------------------
+                  | Retrieve Execution Queue (All Final Tasks)|
+                   -------------------------------------------
+                                    |
+                                    V
+                -------------------------------------------------
+               | Validate and Apply Results from Dependent Tasks |
+                -------------------------------------------------
+                                    | (if the validation fails, declare that task failure but
+                                    | continue executing rest of the final tasks)
+                                    V
+                    ------------------------------------
+                   | Execute the Task and Create TaskRun |
+                    ------------------------------------
+```
+
+## Advantages
+
+* No API changes.
+
+* Finally contract does not break and all final tasks are attempted.
+
+* Consistent across non-final and final tasks.
+
+* Finally tasks do not need to explicitly check if the Task they depend on ran or not
+
+* This proposal is extensible with a param validation proposal using Schema -
+[#1393](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526) by reading default from
+`Task` specification.
+
+> `PipelineRun` reads parameter defaults from `Pipeline` specification but in case of
+parameters associated with `Task` `Results`, the default values will be used from the `Task` since
+`Pipeline` has no direct mapping between `Pipeline` parameter and `Task` parameter.
```

Comment:
is the example below meant to demonstrate that even if `boskos-acquire` fails, `boskos-release` will still run because it has a default for the param? if so i think we might want to say that explicitly

Reply:
yup, added it explicitly 😄